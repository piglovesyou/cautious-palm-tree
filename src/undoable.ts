export class Undoable<CellType> {
  curr;
  max: number
  hist: CellType[][] = [];
  pos: number = 0;
  constructor(data, max) {
    this.hist.push(this.curr = data);
    this.max = max
  }
  setNewValue(rowIndex, columnId, value) {
    const data = this.curr = Object.create(this.curr);
    const cellData = data[rowIndex] = Object.create(data[rowIndex]);
    cellData[columnId] = value;

    this.hist.push(data);
    if (this.hist.length > this.max)
      this.hist.shift();
    else
      this.pos = this.hist.length - 1
    return this.curr;
  }
  undo() {
    if (this.pos <= 0) {
      window.alert(`Unable to undo. You're already in the edge of history.`)
      return this.curr;
    }
    if (this.curr.__proto__ === Array.prototype) throw new Error('Never');
    let nextPos = this.pos - 1;
    if (this.hist[nextPos] !== this.curr.__proto__) throw new Error('Never');
    this.pos = nextPos;
    return this.curr = this.curr.__proto__;
  }
  redo() {
    if (this.pos >= this.max - 1) {
      window.alert(`Unable to redo. You're already in the edge of history.`)
      return this.curr;
    }
    let nextPos = this.pos + 1;
    const newData = this.hist[nextPos];
    if (!newData) throw new Error('Never');
    this.pos = nextPos;
    return this.curr = newData;
  }
  getData() {
    return this.curr;
  }
  toArray() {
    return this.curr.map(e => e);
  }
}
