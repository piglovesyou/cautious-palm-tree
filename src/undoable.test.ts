import { Undoable } from "./undoable";

describe('undoable.ts', () => {
  test('undoable', () => {
    const a0 = [{a: 'a'}, {a: 'a'}, {a: 'a'}];
    const u = new Undoable<{a: string}>(a0, 10);

    u.setNewValue(2, 'a', 'blaa');
    expect(u.toArray()).toStrictEqual([{a: 'a'}, {a: 'a'}, {a: 'blaa'}])

    u.undo();
    expect(u.toArray()).toStrictEqual([{a: 'a'}, {a: 'a'}, {a: 'a'}])

    u.redo();
    expect(u.toArray()).toStrictEqual([{a: 'a'}, {a: 'a'}, {a: 'blaa'}])
  });

  test('deeper', () => {
    let alertMessage: string;
    jest.spyOn(window, 'alert').mockImplementation((m) => {alertMessage = m});

    const a0 = [{a: 'a'}, {a: 'a'}, {a: 'a'}];
    const u = new Undoable<{a: string}>(a0, 10);

    u.setNewValue(1, 'a', 'blaa1');
    u.setNewValue(0, 'a', 'blaa2');
    u.setNewValue(1, 'a', 'blaa3');
    u.setNewValue(2, 'a', 'blaa4');
    u.setNewValue(1, 'a', 'blaa5');
    u.setNewValue(2, 'a', 'blaa6');
    u.setNewValue(1, 'a', 'blaa7');
    u.setNewValue(0, 'a', 'blaa8');
    u.setNewValue(1, 'a', 'blaa9');
    u.setNewValue(1, 'a', 'blaa10');

    expect(alertMessage!).toMatchSnapshot();
    u.redo();
    expect(alertMessage!).toMatchSnapshot();
    alertMessage = ''
    expect(u.toArray()).toMatchSnapshot();

    u.undo();
    u.undo();
    u.undo();
    u.undo();
    u.undo();
    u.undo();
    u.undo();
    u.undo();
    u.undo();

    expect(alertMessage!).toStrictEqual('')
    u.undo();
    expect(alertMessage!).toMatchSnapshot();
    alertMessage = ''
    expect(u.toArray()).toMatchSnapshot();

    u.redo();
    u.redo();
    u.redo();
    u.redo();
    u.redo();
    u.redo();
    u.redo();
    u.redo();
    u.redo();

    expect(alertMessage!).toStrictEqual('')
    u.redo();
    expect(alertMessage!).toMatchSnapshot();
    alertMessage = ''
    expect(u.toArray()).toMatchSnapshot();
  })
});
