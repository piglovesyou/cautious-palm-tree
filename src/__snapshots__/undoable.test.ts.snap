// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`undoable.ts deeper 1`] = `undefined`;

exports[`undoable.ts deeper 2`] = `"Unable to redo. You're already in the edge of history."`;

exports[`undoable.ts deeper 3`] = `
Array [
  Object {
    "a": "blaa8",
  },
  Object {
    "a": "blaa10",
  },
  Object {
    "a": "blaa6",
  },
]
`;

exports[`undoable.ts deeper 4`] = `"Unable to undo. You're already in the edge of history."`;

exports[`undoable.ts deeper 5`] = `
Array [
  Object {
    "a": "a",
  },
  Object {
    "a": "blaa1",
  },
  Object {
    "a": "a",
  },
]
`;

exports[`undoable.ts deeper 6`] = `"Unable to redo. You're already in the edge of history."`;

exports[`undoable.ts deeper 7`] = `
Array [
  Object {
    "a": "blaa8",
  },
  Object {
    "a": "blaa10",
  },
  Object {
    "a": "blaa6",
  },
]
`;
